utils/logger/logger.go

```diff
@@ -0,0 +1,205 @@
+// this is a new logger interface for mattermost
+
+package logger
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"runtime"
+
+	l4g "github.com/alecthomas/log4go"
+
+	"github.com/mattermost/platform/model"
+	"github.com/mattermost/platform/utils"
+)
+
+// this pattern allows us to "mock" the underlying l4g code when unit testing
+var debug = l4g.Debug
+var info = l4g.Info
+var err = l4g.Error
+
+func init() {
+	initL4g(utils.Cfg)
+}
+
+// listens for configuration changes that we might need to respond to
+var configListenerID = utils.AddConfigListener(func(oldConfig *model.Config, newConfig *model.Config) {
+	info("Configuration change detected, reloading log settings")
+	initL4g(newConfig)
+})
+
+// assumes that ../config.go::configureLog has already been called, and has in turn called l4g.close() to clean up
+// any old filters that we might have previously created
+func initL4g(config *model.Config) {
+	// TODO: add support for newConfig.LogSettings.EnableConsole. Right now, ../config.go sets it up in its configureLog
+	// method. If we also set it up here, messages will be written to the console twice. Eventually, when all instances
+	// of l4g have been replaced by this logger, we can move that code to here
+	if config.LogSettings.EnableFile {
+		level := l4g.DEBUG
+		if config.LogSettings.FileLevel == "INFO" {
+			level = l4g.INFO
+		} else if config.LogSettings.FileLevel == "WARN" {
+			level = l4g.WARNING
+		} else if config.LogSettings.FileLevel == "ERROR" {
+			level = l4g.ERROR
+		}
+
+		// create a logger that writes JSON objects to a file, and override our log methods to use it
+		flw := NewJSONFileLogger(level, utils.GetLogFileLocation(config.LogSettings.FileLocation)+".jsonl")
+		debug = flw.Debug
+		info = flw.Info
+		err = flw.Error
+	}
+}
+
+// contextKey lets us add contextual information to log messages
+type contextKey string
+
+func (c contextKey) String() string {
+	return string(c)
+}
+
+const contextKeyUserID contextKey = contextKey("user-id")
+const contextKeyRequestID contextKey = contextKey("request-id")
+
+// any contextKeys added to this array will be serialized in every log message
+var contextKeys = [2]contextKey{contextKeyUserID, contextKeyRequestID}
+
+// WithUserID adds a user id to the specified context. If the returned Context is subsequently passed to a logging
+// method, the user id will automatically be included in the logged message
+func WithUserID(ctx context.Context, userID string) context.Context {
+	return context.WithValue(ctx, contextKeyUserID, userID)
+}
```

Comment:
Well... you know how I feel about using context for this purpose to begin with. (I'm pretty against it.)

I'll try not to beat a dead horse, but I will say that I sincerely hope context is removed or reworked in Go 2.0, whenever that day comes. Which I'm optimistic about because that value-storing interface is one of the biggest sources of contention in the standard library and even the language designers don't know how it should be used (see [Russ's new years resolutions](https://research.swtch.com/go2017)).